R/driver_rds_s3.R

```diff
@@ -0,0 +1,335 @@
+      }
+      self$mangle_key <- driver_rds_s3_config(bucket, path, "mangle_key", mangle_key,
+                                              FALSE, TRUE)
+
+      if (!is.null(mangle_key_pad)) {
+        assert_scalar_logical(mangle_key_pad)
+      }
+      self$mangle_key_pad <-
+        driver_rds_s3_config(bucket, path, "mangle_key_pad", mangle_key_pad,
+                             FALSE, TRUE)
+
+      if (!is.null(compress)) {
+        assert_scalar_logical(compress)
+      }
+      self$compress <- driver_rds_s3_config(bucket, path, "compress", compress,
+                                            TRUE, FALSE)
+
+      if (!is.null(hash_algorithm)) {
+        assert_scalar_character(hash_algorithm)
+      }
+      self$hash_algorithm <- driver_rds_s3_config(bucket, path, "hash_algorithm",
+                                                  hash_algorithm, "md5", TRUE)
+    },
+
+    type = function() {
+      "rds_s3"
+    },
+    destroy = function() {
+      s3_delete_recursive(bucket = self$bucket, path = self$path)
+    },
+
+    get_hash = function(key, namespace) {
+      s3_readLines(path = self$name_key(key, namespace), bucket = self$bucket)
+    },
+    set_hash = function(key, namespace, hash) {
+      dir_create(self$name_key("", namespace))
+      s3_writeLines(text = hash, path = self$name_key(key, namespace), bucket = self$bucket) #*** should be making use of (or making an equivalent version of) the write_lines function within the storr package here (I think it deletes file if the write fails)
+    },
+    get_object = function(hash) {
+      aws.s3::s3readRDS(object = self$name_hash(hash), bucket = self$bucket)
+    },
+    set_object = function(hash, value) {
+      ## NOTE: this takes advantage of having the serialized value
+      ## already and avoids seralising twice.
+      assert_raw(value)
+      aws.s3::s3write_using(x = value, FUN = function(v, p) write_serialized_rds(v, p, self$compress), object = self$name_hash(hash), bucket = self$bucket)
+    },
+
+    exists_hash = function(key, namespace) {
+      s3_object_exists(self$name_key(key, namespace), bucket = self$bucket)
+    },
+    exists_object = function(hash) {
+      s3_object_exists(self$name_hash(hash), bucket = self$bucket)
+    },
+
+    del_hash = function(key, namespace) {
+      #s3_delete_file(bucket = self$bucket, path = self$name_key(key, namespace))
+      ## above deletes just one file (s3 key).
+      ## However it will throw an error if the file we are trying to delete looks like a directory.
+      ## S3 has no actual notion of directory, we just fake it using "/". As a result, it's possible to get into a muddle.
+      ## to play it save, line below can be uncommented to force it to delete just the path given, but throw a warning, if it does look like a directory
+      ## can also change to if_dir = "del_recursive" to delete the whole directory with a warning. May never actually show up as an issue, this is just a note.
```

Comment:
If we're the only modifying the bucket is this a real concern?  I don't know enough about S3!  Can you reflow to 80 characters please?